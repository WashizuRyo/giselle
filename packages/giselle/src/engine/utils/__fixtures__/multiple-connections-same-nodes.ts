/**
 * Multiple Connections Between Same Nodes Test Fixture
 *
 * This fixture tests the scenario where one node has multiple outputs
 * connected to the same target node - a common pattern when passing
 * multiple parameters from a trigger to an action.
 *
 * Node Structure:
 *
 * ┌─────────────────────────────┐
 * │ On Issue Comment Created    │
 * │ (GitHub Trigger)            │
 * │ nd-nwdgrtVwC070isLr         │
 * └──┬──────────────┬───────────┘
 *    │              │
 *    │ Issue Number │ Issue Title
 *    │              │
 *    ▼              ▼
 * ┌─────────────────────────────┐
 * │ Create Issue Comment        │
 * │ (GitHub Action)             │
 * │ nd-PlnxfHCFVLCuVJcb         │
 * └─────────────────────────────┘
 *
 * Key Points:
 * - Trigger has 4 outputs: issueComment, issueNumber, issueTitle, issueBody
 * - Action has 2 inputs: issueNumber, body
 * - TWO connections from same trigger to same action:
 *   1. issueNumber (otp-dpz92nyiAVTcoLs6) → issueNumber (inp-rKOQKGqGb7Qeps2X)
 *   2. issueTitle (otp-CTj1biBXwjzZuqNo) → body (inp-FiinOx0HrQqamro9)
 *
 * This tests the bug fix where filter() instead of find() ensures ALL
 * connections between the same two nodes are collected.
 */

import type { Workspace } from "@giselle-sdk/data-type";

export const multipleConnectionsSameNodesFixture: Workspace = {
	id: "wrks-JJqRLOtbigJ6Tc58",
	schemaVersion: "20250221",
	nodes: [
		{
			id: "nd-nwdgrtVwC070isLr",
			name: "On Issue Comment Created",
			type: "operation",
			inputs: [],
			outputs: [
				{
					id: "otp-QAq4xmw2KdTuLgP2",
					label: "Issue Comment",
					accessor: "body",
				},
				{
					id: "otp-dpz92nyiAVTcoLs6",
					label: "Issue Number",
					accessor: "issueNumber",
				},
				{
					id: "otp-CTj1biBXwjzZuqNo",
					label: "Issue Title",
					accessor: "issueTitle",
				},
				{
					id: "otp-Mhi9HxfxtZQ6fcTt",
					label: "Issue Body",
					accessor: "issueBody",
				},
			],
			content: {
				type: "trigger",
				provider: "github",
				state: {
					status: "configured",
					flowTriggerId: "fltg-ZtwjC12CVW5kPdUu",
				},
			},
		},
		{
			id: "nd-PlnxfHCFVLCuVJcb",
			name: "Create Issue Comment",
			type: "operation",
			inputs: [
				{
					id: "inp-rKOQKGqGb7Qeps2X",
					label: "issueNumber",
					accessor: "issueNumber",
					isRequired: true,
				},
				{
					id: "inp-FiinOx0HrQqamro9",
					label: "body",
					accessor: "body",
					isRequired: true,
				},
			],
			outputs: [
				{
					id: "otp-FZaotlnouWEScbaJ",
					label: "output",
					accessor: "action-result",
				},
			],
			content: {
				type: "action",
				command: {
					provider: "github",
					state: {
						status: "configured",
						commandId: "github.create.issueComment",
						repositoryNodeId: "R_kgDON5_TWw",
						installationId: 81253657,
					},
				},
			},
		},
	],
	connections: [
		{
			id: "cnnc-mzeOjkqocn9YFhlD",
			outputNode: {
				id: "nd-nwdgrtVwC070isLr",
				type: "operation",
				content: {
					type: "trigger",
				},
			},
			outputId: "otp-dpz92nyiAVTcoLs6",
			inputNode: {
				id: "nd-PlnxfHCFVLCuVJcb",
				type: "operation",
				content: {
					type: "action",
				},
			},
			inputId: "inp-rKOQKGqGb7Qeps2X",
		},
		{
			id: "cnnc-pf2KwXJAoEEBCE1n",
			outputNode: {
				id: "nd-nwdgrtVwC070isLr",
				type: "operation",
				content: {
					type: "trigger",
				},
			},
			outputId: "otp-CTj1biBXwjzZuqNo",
			inputNode: {
				id: "nd-PlnxfHCFVLCuVJcb",
				type: "operation",
				content: {
					type: "action",
				},
			},
			inputId: "inp-FiinOx0HrQqamro9",
		},
	],
	ui: {
		nodeState: {
			"nd-nwdgrtVwC070isLr": {
				position: {
					x: -199.3909277921317,
					y: 1051.8414322662375,
				},
				selected: true,
				showError: false,
				highlighted: false,
				measured: {
					width: 260,
					height: 260,
				},
			},
			"nd-PlnxfHCFVLCuVJcb": {
				position: {
					x: 310.4926866012552,
					y: 1063.047396001855,
				},
				selected: false,
				showError: false,
				highlighted: false,
				measured: {
					width: 229,
					height: 195,
				},
			},
		},
		viewport: {
			x: 543.0285601454921,
			y: -789.8216993868937,
			zoom: 0.9113626570467636,
		},
		currentShortcutScope: "canvas",
		selectedConnectionIds: [],
	},
};
