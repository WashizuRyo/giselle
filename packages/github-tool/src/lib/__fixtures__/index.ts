import { readFileSync } from "node:fs";
import { join } from "node:path";

function loadFixture(filename: string): string {
	const fixturePath = join(__dirname, filename);
	return readFileSync(fixturePath, "utf-8");
}

export const PR_1118_DIFF = loadFixture("pr-1118-diff.txt");
