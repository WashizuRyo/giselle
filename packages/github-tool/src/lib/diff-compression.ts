export interface FileDiff {
	fileName: string;
	content: string;
	size: number;
	header: string;
}

export function compressLargeDiff(diff: string, maxSize = 8000): string {
	if (diff.length <= maxSize) {
		return diff;
	}

	const fileDiffs = parseFileDiffs(diff);

	fileDiffs.sort((a, b) => b.size - a.size);
	const totalHeaderSize = fileDiffs.reduce((sum, file) => {
		const headerSize = file.content
			.split("\n")
			.filter((line) => isHeaderLine(line))
			.join("\n").length;
		return sum + headerSize;
	}, 0);

	const availableForContent = maxSize - totalHeaderSize;
	const totalContentSize =
		fileDiffs.reduce((sum, file) => sum + file.size, 0) - totalHeaderSize;
	const compressionRatio = Math.min(1, availableForContent / totalContentSize);

	const compressedFiles = fileDiffs.map((file) => {
		const headerSize = file.content
			.split("\n")
			.filter((line) => isHeaderLine(line))
			.join("\n").length;
		const contentSize = file.size - headerSize;
		const targetContentSize = Math.floor(contentSize * compressionRatio);
		const targetTotalSize = headerSize + targetContentSize;

		return compressFileDiff(file, targetTotalSize);
	});

	return compressedFiles.join("\n");
}

function parseFileDiffs(diff: string): FileDiff[] {
	const lines = diff.split("\n");
	const fileDiffs: FileDiff[] = [];
	let currentFile: FileDiff | null = null;

	for (const line of lines) {
		if (line.startsWith("diff --git")) {
			if (currentFile) {
				fileDiffs.push(currentFile);
			}

			const fileName = extractFileName(line);
			currentFile = {
				fileName,
				content: `${line}\n`,
				size: line.length + 1,
				header: line,
			};
		} else if (currentFile) {
			currentFile.content += `${line}\n`;
			currentFile.size += line.length + 1;
		}
	}

	if (currentFile) {
		fileDiffs.push(currentFile);
	}

	return fileDiffs;
}

function compressFileDiff(fileDiff: FileDiff, targetSize: number): string {
	if (fileDiff.size <= targetSize) {
		return fileDiff.content.trim();
	}

	const lines = fileDiff.content.split("\n");
	let result = "";
	let currentSize = 0;

	const headerLines: string[] = [];
	const contentLines: string[] = [];

	for (const line of lines) {
		if (isHeaderLine(line)) {
			headerLines.push(line);
		} else {
			contentLines.push(line);
		}
	}

	const headerContent = `${headerLines.join("\n")}\n`;
	result = headerContent;
	currentSize = headerContent.length;

	const remainingSize = targetSize - currentSize - 4;
	let contentAdded = "";

	for (const line of contentLines) {
		const lineWithNewline = `${line}\n`;
		if (currentSize + lineWithNewline.length <= remainingSize) {
			contentAdded += lineWithNewline;
			currentSize += lineWithNewline.length;
		} else {
			break;
		}
	}

	result += contentAdded;

	if (contentLines.length > contentAdded.split("\n").length - 1) {
		result += "...\n";
	}

	return result.trim();
}

function isHeaderLine(line: string): boolean {
	return (
		line.startsWith("diff --git") ||
		line.startsWith("index ") ||
		line.startsWith("+++") ||
		line.startsWith("---") ||
		line.startsWith("@@") ||
		line.match(/^new file mode/) !== null ||
		line.match(/^deleted file mode/) !== null
	);
}

function extractFileName(diffLine: string): string {
	const match = diffLine.match(/diff --git a\/(.+?) b\//);
	return match ? match[1] : "unknown";
}
