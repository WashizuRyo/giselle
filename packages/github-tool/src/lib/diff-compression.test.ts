import { describe, expect, it } from "vitest";
import { PR_1118_DIFF } from "./__fixtures__";
import { type FileDiff, compressLargeDiff } from "./diff-compression";

describe("diff compression", () => {
	it("should return original diff when under maxSize", () => {
		const smallDiff = "diff --git a/test.txt b/test.txt\n+hello world";
		const result = compressLargeDiff(smallDiff, 1000);
		expect(result).toBe(smallDiff);
	});

	it("should compress PR #1118 diff to 1500 characters", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 1500);

		expect(compressed.length).toBeLessThanOrEqual(1500);
		expect(compressed).toContain("diff --git");
		expect(compressed).toContain("ingest-github-repository.ts");
		expect(compressed).toContain("route.ts");
	});

	it("should preserve file headers in compressed output", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 1500);

		expect(compressed).toContain(
			"diff --git a/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts",
		);
		expect(compressed).toContain("new file mode 100644");
		expect(compressed).toContain("index 000000000..b5ad8bde7");
		expect(compressed).toContain("--- /dev/null");
		expect(compressed).toContain(
			"+++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts",
		);
	});

	it("should add truncation indicator when content is cut", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 1500);

		expect(compressed).toContain("...");
	});

	it("should handle empty diff", () => {
		const result = compressLargeDiff("", 1000);
		expect(result).toBe("");
	});

	it("should handle single file diff", () => {
		const singleFileDiff = `diff --git a/test.txt b/test.txt
index 1234567..abcdefg 100644
--- a/test.txt
+++ b/test.txt
@@ -1,3 +1,3 @@
 line 1
-line 2
+line 2 modified
 line 3`;

		const compressed = compressLargeDiff(singleFileDiff, 150);
		expect(compressed.length).toBeLessThanOrEqual(150);
		expect(compressed).toContain("diff --git");
		expect(compressed).toContain("test.txt");
	});

	it("should show actual compression ratio for PR #1118", () => {
		const originalSize = PR_1118_DIFF.length;
		const compressed = compressLargeDiff(PR_1118_DIFF, 1500);
		const compressedSize = compressed.length;
		const compressionRatio = (
			((originalSize - compressedSize) / originalSize) *
			100
		).toFixed(1);

		console.log(`Original size: ${originalSize} characters`);
		console.log(`Compressed size: ${compressedSize} characters`);
		console.log(`Compression ratio: ${compressionRatio}%`);
		console.log("First 200 chars of compressed output:");
		console.log(`${compressed.substring(0, 200)}...`);

		expect(compressedSize).toBeLessThanOrEqual(1500);
		expect(originalSize).toBeGreaterThan(compressedSize);
	});

	it("should match snapshot for PR #1118 at 1500 chars", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 1500);
		expect(compressed).toMatchSnapshot("pr-1118-compressed-1500.txt");
	});

	it("should match snapshot for PR #1118 at 2000 chars", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 2000);
		expect(compressed).toMatchSnapshot("pr-1118-compressed-2000.txt");
	});

	it("should match snapshot for PR #1118 at 1000 chars", () => {
		const compressed = compressLargeDiff(PR_1118_DIFF, 1000);
		expect(compressed).toMatchSnapshot("pr-1118-compressed-1000.txt");
	});
});
