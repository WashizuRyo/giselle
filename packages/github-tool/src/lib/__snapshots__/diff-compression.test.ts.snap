// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`diff compression > should match snapshot for PR #1118 at 1000 chars > pr-1118-compressed-1000.txt 1`] = `
"diff --git a/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
new file mode 100644
index 000000000..b5ad8bde7
--- /dev/null
+++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
@@ -0,0 +1,90 @@
++++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/route.ts
+import { db, githubRepositoryIndex } from "@/drizzle";
+import {
+	type GitHubChunkMetadata,
..."
`;

exports[`diff compression > should match snapshot for PR #1118 at 1500 chars > pr-1118-compressed-1500.txt 1`] = `
"diff --git a/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
new file mode 100644
index 000000000..b5ad8bde7
--- /dev/null
+++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
@@ -0,0 +1,90 @@
++++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/route.ts
+import { db, githubRepositoryIndex } from "@/drizzle";
+import {
+	type GitHubChunkMetadata,
+	createGitHubChunkStore,
+} from "@/lib/vector-stores/github-blob-stores";
+import {
+	GitHubBlobLoader,
+	type GitHubBlobLoaderParams,
+	type GitHubBlobMetadata,
+} from "@giselle-sdk/github-tool";
+import { createIngestPipeline } from "@giselle-sdk/rag2";
+import type { Octokit } from "@octokit/core";
+import { and, eq } from "drizzle-orm";
+
+/**
+ * Main GitHub repository ingestion coordination
+ */
+export async function ingestGitHubRepository(params: {
+	octokitClient: Octokit;
..."
`;

exports[`diff compression > should match snapshot for PR #1118 at 2000 chars > pr-1118-compressed-2000.txt 1`] = `
"diff --git a/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
new file mode 100644
index 000000000..b5ad8bde7
--- /dev/null
+++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/ingest-github-repository.ts
@@ -0,0 +1,90 @@
++++ b/apps/studio.giselles.ai/app/api/vector-stores/github/ingest/route.ts
+import { db, githubRepositoryIndex } from "@/drizzle";
+import {
+	type GitHubChunkMetadata,
+	createGitHubChunkStore,
+} from "@/lib/vector-stores/github-blob-stores";
+import {
+	GitHubBlobLoader,
+	type GitHubBlobLoaderParams,
+	type GitHubBlobMetadata,
+} from "@giselle-sdk/github-tool";
+import { createIngestPipeline } from "@giselle-sdk/rag2";
+import type { Octokit } from "@octokit/core";
+import { and, eq } from "drizzle-orm";
+
+/**
+ * Main GitHub repository ingestion coordination
+ */
+export async function ingestGitHubRepository(params: {
+	octokitClient: Octokit;
+	source: { owner: string; repo: string; commitSha: string };
+	teamDbId: number;
+}): Promise<void> {
+	const repositoryIndexDbId = await getRepositoryIndexDbId(
+		params.source,
+		params.teamDbId,
+	);
+
+	const githubLoader = new GitHubBlobLoader(params.octokitClient, {
+		maxBlobSize: 1 * 1024 * 1024,
+	});
+	const chunkStore = createGitHubChunkStore(repositoryIndexDbId);
+
+	const pipeline = createIngestPipeline<
+		GitHubBlobMetadata,
+		GitHubChunkMetadata,
+		GitHubBlobLoaderParams
+	>({
..."
`;
